Add assert_coltypes

FlukeAndFeather opened the pull request:
What do you think for this approach for assert_coltypes? [Your comment](https://github.com/gnguy/assertable/issues/3#issuecomment-530565150) in issue #3 suggests taking e.g. `is.numeric` and this version takes a vector as a template e.g. `double(0)`. I think this way makes more sense for checking *just* column types and your suggestion works better for general purpose column assertions that take any function. For example something like this:

```
assertthat_col <- function(data, predicates, quiet = FALSE) {
  # Do all specified colnames exist in data frame?
  non_df_cols <- colnames(data)[!(names(predicates) %in% colnames(data))]
  if(length(non_df_cols) > 0) {
    stop(paste0("These columns exist in colnames but not in your dataframe: ",
                paste(non_df_cols, collapse=" ")))
  }
  
  # Test if columns pass their predicates
  colpass <- mapply(function(.f, .col) .f(data[, .col]), predicates, names(predicates))
  if(any(!colpass)) {
    stop(paste0("These columns in your data frame failed their predicates: ",
                paste(names(predicates)[!colpass], collapse=" ")))
  }
  
  if(!quiet) {
    print("All columns pass their predicates")
  }
}

# In data frame CO2, assert that mean of column `conc` is less than 500 and `uptake` is positive.
preds_pass <- list(uptake = function(col) mean(col) <= 500,
                   conc = function(col) all(col > 0))
preds_fail <- list(uptake = function(col) mean(col) > 500,
                   conc = function(col) all(col < 0))
assertthat_col(CO2, preds_pass)
assertthat_col(CO2, preds_fail)
```

What do you think? I don't love the name `assertthat_col` but I think the general idea could work.

Commit:
Add assert_coltypes

Also adds me to DESCRIPTION as a contributor

Commit:
Add assert_coltypes

Also adds me to DESCRIPTION as a contributor

gnguy reviewed (changes requested):
Thanks for these changes! Here are some pieces of feedback:

- Line 18: Needs a closing parenthesis on the example.
- Lines 16 and 18: What are your thoughts about removing the 0's from the examples? e.g. updating to `assert_coltypes(CO2, list(Plant = integer(), conc = double()))` -- the 0's tripped me up at first as I was trying to figure out why the 0 was important/required to have, but it seems like 0 is the default arg for these so the more minimal example also works.
- Lines 29-33: Neat implementation! Would it be possible to update the error message to something like this example below? Ideally it'd be nice for the error message to return not only the column that failed but also the correct coltype, to skip that debugging step.

```
  if(any(!typematch)) {
    error_cols <- names(coltypes)[!typematch]
    message_list <- lapply(error_cols, function(x) return(paste0(x, ": ", typeof(data[[x]]))))
    stop(paste0("These columns in your data frame have different types than supplied in coltypes (actual coltypes listed): ",
                paste(message_list, collapse = ", ")))
  }
```

Commit:
Fix assert_coltypes examples and add a more verbose error message

Commit:
Merge branch 'coltypes' of https://github.com/FlukeAndFeather/assertable into coltypes

FlukeAndFeather commented:
I'm still new to using GitHub forking workflows, so I apologize in advance for any missteps! I addressed the issues in your review and pushed them to the same branch. I think they should be available in this same pull request, right? Apologies for the messy commit history; I followed a set of instructions that recommended rebasing.

gnguy reviewed (approved):
Looks great, thanks for contributing!
